tensorboard/plugins/core/core_plugin.py

```diff
@@ -115,6 +125,29 @@ def _serve_asset(self, path, gzipped_asset_bytes, request):
             request, gzipped_asset_bytes, mimetype, content_encoding="gzip"
         )
 
+    @wrappers.Request.application
+    def _serve_index(self, index_asset_bytes, request):
+        """Serves index.html content.
+
+        Note that we opt out of gzipping index.html to write preamble before the
+        resource content. This inflates the resource size from 2x kiB to 1xx
+        kiB, but we require an ability to flush preamble with the HTML content.
+        """
+        relpath = (
+            posixpath.relpath(self._path_prefix, request.script_root)
+            if self._path_prefix
+            else "."
+        )
+        # Technically, it is possible to flush parts using yields but http_utils
+        # try to measure gzip and measure content length if a truthy value is
+        # passed.
+        res = http_util.Respond(request, None, "text/html")
+        res.stream.write(
+            '<meta name="tb-relative-root" content="%s/" />' % (relpath)
+        )
+        res.stream.write(index_asset_bytes)
+        return res
```

Comment:
I don’t quite understand why this change implies the accompanying change
to `http_util.Respond`. The alternative that comes to mind is:

```python
meta_header = '<meta name="tb-relative-root" content="%s/" />' % (relpath)
content = meta_header.encode("utf-8") + index_asset_bytes
return http_util.Respond(content)
```

Judging from the comment, it sounds like you want to avoid that because
it automatically gzips, since the content type is `text/html`. That
makes sense. But then can’t we just pass `content_encoding="identity"`
rather than adding new mechanisms? I think that that already does the
right thing:

```python
>>> from tensorboard.backend import http_util
>>> from werkzeug import test as wtest
>>> from werkzeug import wrappers
>>> def req():
...     headers = [("Accept-Encoding", "gzip")]
...     environ = wtest.EnvironBuilder(headers=headers).get_environ()
...     return wrappers.Request(environ)
... 
>>> res = http_util.Respond(req(), b"<html>\n", "text/html")
>>> res.headers.get("Content-Encoding"), res.get_data()
('gzip', b'\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\xff\xb3\xc9(\xc9\xcd\xb1\xe3\x02\x00\x994\xcb3\x07\x00\x00\x00')
>>> res = http_util.Respond(req(), b"<html>\n", "text/html", content_encoding="identity")
>>> res.headers.get("Content-Encoding"), res.get_data()
```

Failing that, I might prefer to add a new `no_gzip` flag or something
rather than introducing (a) the edge case in `http_util` and (b) the
stream-writing code in this method.